controllers/keystoneapi_controller.go

```diff
@@ -250,6 +255,72 @@ func (r *KeystoneAPIReconciler) reconcileInit(ctx context.Context, instance *key
 
 	// run keystone db sync - end
 
+	//
+	// expose the service (create service, route and return the created endpoint URLs)
+	//
+	var keystonePorts = map[string]int32{
+		"admin":    keystone.KeystoneAdminPort,
+		"public":   keystone.KeystonePublicPort,
+		"internal": keystone.KeystoneInternalPort,
+	}
+
+	apiEndpoints, ctrlResult, err := common.ExposeEndpoints(
+		ctx,
+		helper,
+		keystone.ServiceName,
+		endpointLabels,
+		keystonePorts,
+	)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+
+	//
+	// Update instance status with service endpoint url from route host information
+	//
+	// TODO: need to support https default here
+	if instance.Status.APIEndpoints == nil {
+		instance.Status.APIEndpoints = map[string]string{}
+	}
+	instance.Status.APIEndpoints = apiEndpoints
+
+	// expose service - end
+
+	//
+	// BootStrap Job
+	//
+	jobDef = keystone.BootstrapJob(instance, instance.Status.APIEndpoints)
+	bootstrapjob := common.NewJob(
+		jobDef,
+		keystonev1.BootstrapHash,
+		instance.Spec.PreserveJobs,
+		5,
+		instance.Status.Hash[keystonev1.BootstrapHash],
+	)
+	ctrlResult, err = bootstrapjob.DoJob(
+		ctx,
+		helper,
+	)
+	if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
Should we use the same pattern here that we've used elsewhere, where we check for `err != nil` first?  I don't know if it really matters, but it's just something I noticed

Reply:
yes I think we could change it again. how the function was implemented to wait for the job to finish required this order to not print errors when e.g. the job was not yet created/available. but I changed this particular case. I'll retest/recheck it and may change it in a follow up.